application/src/main/java/org/opentripplanner/graph_builder/module/osm/WalkableAreaBuilder.java

```diff
@@ -185,49 +168,41 @@ public void buildWithVisibility(AreaGroup group) {
     // List of edges belonging to the walkable area
     Set<Edge> edges = new HashSet<>();
 
-    // Edges which are part of the rings. We want to keep there for linking even tough they
-    // might not be part of the visibility edges.
+    // Edges which are part of the rings
     Set<Edge> ringEdges = new HashSet<>();
 
     // OSM ways that this area group consists of
     Set<Long> osmWayIds = group.areas
       .stream()
       .map(area -> area.parent)
-      .flatMap(osmWithTags ->
-        osmWithTags instanceof OsmRelation
-          ? ((OsmRelation) osmWithTags).getMembers().stream().map(OsmRelationMember::getRef)
-          : Stream.of(osmWithTags.getId())
+      .flatMap(osmEntity ->
+        osmEntity instanceof OsmRelation
+          ? ((OsmRelation) osmEntity).getMembers().stream().map(OsmRelationMember::getRef)
+          : Stream.of(osmEntity.getId())
       )
       .collect(Collectors.toSet());
 
-    var references = getStopReferences(group);
-
     // create polygon and accumulate nodes for area
     for (Ring ring : group.outermostRings) {
       Polygon polygon = ring.jtsPolygon;
 
-      AreaEdgeList edgeList = new AreaEdgeList(polygon, references);
+      AreaGroup areaGroup = new AreaGroup(polygon);
 
-      // the points corresponding to concave or hole vertices
-      // or those linked to ways
+      // the points corresponding to concave or hole vertices or those linked to ways
       HashSet<OsmNode> visibilityNodes = new HashSet<>();
       HashSet<NodeEdge> alreadyAddedEdges = new HashSet<>();
       HashSet<IntersectionVertex> platformLinkingVertices = new HashSet<>();
-      // we need to accumulate visibility points from all contained areas
-      // inside this ring, but only for shared nodes; we don't care about
-      // convexity, which we'll handle for the grouped area only.
-
       GeometryFactory geometryFactory = GeometryUtils.getGeometryFactory();
 
-      OsmWithTags areaEntity = group.getSomeOsmObject();
+      OsmEntity areaEntity = group.getSomeOsmObject();
 
-      // we also want to fill in the edges of this area anyway, because we can,
-      // and to avoid the numerical problems that they tend to cause
-      for (Area area : group.areas) {
-        if (!polygon.contains(area.jtsMultiPolygon)) {
-          continue;
+      for (OsmArea area : group.areas) {
+        // test if area is inside current ring. This is necessary only if there are many areas or outer rings
+        if (group.areas.size() != 1 || group.outermostRings.size() != 1) {
```

Comment:
Can you introduce a method on the group that answers that question. Then you can put the line comment on it as Javadoc.